storm-server/src/main/java/org/apache/storm/scheduler/blacklist/BlacklistScheduler.java

```diff
@@ -0,0 +1,248 @@
+                blacklistHostSet.add(host);
+            } else {
+                LOG.info("supervisor {} is not alive, do not need to add to blacklist.", supervisor);
+            }
+        }
+        return blacklistHostSet;
+    }
+
+    /**
+     * supervisor or port never exits once in tolerance time will be removed from cache.
+     */
+    private void removeLongTimeDisappearFromCache() {
+
+        Map<String, Integer> supervisorCountMap = new HashMap<String, Integer>();
+        Map<WorkerSlot, Integer> slotCountMap = new HashMap<WorkerSlot, Integer>();
+
+        for (Map<String, Set<Integer>> item : badSupervisorsToleranceSlidingWindow) {
+            Set<String> supervisors = item.keySet();
+            for (String supervisor : supervisors) {
+                int supervisorCount = supervisorCountMap.getOrDefault(supervisor, 0);
+                Set<Integer> slots = item.get(supervisor);
+                if (slots.equals(cachedSupervisors.get(supervisor))) { // treat supervisor as bad only if all slots are bad
+                    supervisorCountMap.put(supervisor, supervisorCount + 1);
+                }
+                for (Integer slot : slots) {
+                    WorkerSlot workerSlot = new WorkerSlot(supervisor, slot);
+                    int slotCount = slotCountMap.getOrDefault(workerSlot, 0);
+                    slotCountMap.put(workerSlot, slotCount + 1);
+                }
+            }
+        }
+
+        for (Map.Entry<String, Integer> entry : supervisorCountMap.entrySet()) {
+            String key = entry.getKey();
+            int value = entry.getValue();
+            if (value == windowSize) { // supervisor which was never back to normal in tolerance period will be removed from cache
+                cachedSupervisors.remove(key);
+                LOG.info("Supervisor {} was never back to normal during tolerance period, probably dead. Will remove from cache.", key);
+            }
+        }
+
+        for (Map.Entry<WorkerSlot, Integer> entry : slotCountMap.entrySet()) {
+            WorkerSlot workerSlot = entry.getKey();
+            String supervisorKey = workerSlot.getNodeId();
+            Integer slot = workerSlot.getPort();
+            int value = entry.getValue();
+            if (value == windowSize) { // worker slot which was never back to normal in tolerance period will be removed from cache
+                Set<Integer> slots = cachedSupervisors.get(supervisorKey);
+                if (slots != null) { // slots will be null while supervisor has been removed from cached supervisors
+                    slots.remove(slot);
+                    cachedSupervisors.put(supervisorKey, slots);
+                }
+                LOG.info("Worker slot {} was never back to normal during tolerance period, probably dead. Will be removed from cache.", workerSlot);
+            }
+        }
+    }
+
+    private Object initializeInstance(String className, String representation) {
```

Comment:
I think there is a function in Utils, or ServerUtils that can do this for us.

Reply:
The refactoring is mainly intended for removing duplicated long `catch` statements. Btw I'll change below line to use Utils.